subsys/bluetooth/host/gatt.c

```diff
@@ -657,6 +680,13 @@ ssize_t bt_gatt_attr_write_ccc(struct bt_conn *conn,
 	/* Update cfg if don't match */
 	if (ccc->cfg[i].value != ccc->value) {
 		gatt_ccc_changed(attr, ccc);
+
+		if (IS_ENABLED(CONFIG_BT_SETTINGS_CCC_STORE_ON_WRITE) &&
+		    bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
+			gatt_ccc_store.conn = conn;
+			k_delayed_work_submit(&gatt_ccc_store.work,
+				CCC_STORE_DELAY);
```

Comment:
k_delayed_work_submit actually cancel and restart its timer if already pending which I guess is not what you want here. Furthermore if the device disconnects while the delayed work is pending you should probably cancel with k_delayed_work_cancel since that would just cause flash access for no reason.

Reply:
@Vudentz fixed.

Reply:
There is flag for that already, iirc k_delayed_work_pending or something like it.

Reply:
'k_work_pending()' checks if the work item is already pending in the queue. But before the timeout of `k_delayed_work` runs out, the item is not pending yet (not yet added to the workqueue) - https://docs.zephyrproject.org/1.7.0/api/kernel_api.html#_CPPv221k_delayed_work_submitP14k_delayed_work7int32_t.

So we either need a custom flag as I did it or we can the fact fact that the work will be canceled and timer restarted.

Reply:
Done. I now cancel the work on disconnection. Also renamed "pending" to "scheduled" as it represents work during timeout, not `pending` in workqueue sense.